x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SparseScoredTermsMapper.java

```diff
@@ -0,0 +1,288 @@
+        new ConstructingObjectParser<>(
+            "sparse_terms_mapped_field",
+            false,
+            a -> new SparseTermsAndScores((List<String>) a[0], (List<Integer>) a[1])
+        );
+    static {
+        SPARSE_TERMS_MAPPED_FIELD_PARSER.declareStringArray(ConstructingObjectParser.constructorArg(), new ParseField("terms"));
+        SPARSE_TERMS_MAPPED_FIELD_PARSER.declareIntArray(ConstructingObjectParser.constructorArg(), new ParseField("scores"));
+    }
+
+    @Override
+    protected void parseCreateField(DocumentParserContext context) throws IOException {
+        context.path().add(simpleName());
+        XContentParser.Token token;
+        XContentSubParser subParser = null;
+        try {
+            token = context.parser().currentToken();
+            if (token == XContentParser.Token.VALUE_NULL) {
+                context.path().remove();
+                return;
+            }
+            // should be an object
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, token, context.parser());
+            subParser = new XContentSubParser(context.parser());
+            SparseTermsAndScores doc = SPARSE_TERMS_MAPPED_FIELD_PARSER.apply(subParser, null);
+            if (doc.terms.size() != doc.scores.size()) {
+                throw new MapperParsingException(
+                    Strings.format(
+                        "failed to parse field [%s] of type [%s], [terms] and [scores] must have equal length; found [%d] and [%d]",
+                        fieldType().name(),
+                        CONTENT_TYPE,
+                        doc.terms.size(),
+                        doc.scores.size()
+                    )
+                );
+            }
+            int i = 0;
+            for (Integer score : doc.scores) {
+                if (score < 0) {
+                    throw new MapperParsingException(
+                        Strings.format(
+                            "failed to parse field [%s] of type [%s], [scores] must be non-negative; found [%d] at position [%d]",
+                            fieldType().name(),
+                            CONTENT_TYPE,
+                            score,
+                            i
+                        )
+                    );
+                }
+                ++i;
+            }
+            if (doc.terms.size() > Sets.newHashSet(doc.terms).size()) {
+                throw new MapperParsingException(
+                    Strings.format(
+                        "failed to parse field [%s] of type [%s]; requires unique items in [terms]",
+                        fieldType().name(),
+                        CONTENT_TYPE
+                    )
+                );
+            }
+            context.doc()
+                .addWithKey(fieldType().name(), new Field(name(), new TermCountStream(doc.terms, doc.scores), Defaults.FIELD_TYPE));
```

Comment:
Does this need to be `addWithKey`? We don't use `getByKey` anywhere, so you should be able to just use `add` here.